### PR TITLE
api: move generic matcher proto to its own package

### PR DESCRIPTION
**Pull request by snowp**
In order to unblock a dependency between route_components.proto and the matcher tree, move the alpha Matcher to 
its own package. To provide an upgrade path for users using this, we'll keep the old copy around for a single deprecation
cycle. 

Signed-off-by: Snow Pettersen <snowp@lyft.com>

Risk Level: Low
Testing: n/a
Docs Changes: n/a 
Release Notes: Deprecation note

### CODE REVIEW

**Comment by snowp** [comment]
@htuch This moves the proto as per #17037. I figured we'd provide some kind of upgrade path via supporting both the new and old proto, though let me know if you think we should just do an aggressive move given the alpha status.

**Review by htuch — COMMENTED** [review]
@snowp very exciting to see routing and sub-linear matchers coming together.

One question I have on this is whether we are breaking the circular dependency at the ideal place. The fact that generic matchers depend on HeaderMatcher in router seems wrong, ideally the header matcher would be somewhere generic. OTOH to fix this would require wasting time deprecating the existing location and moving the type for route table, which is very disruptive to existing users (almost everyone using Envoy and HTTP).

Is what you ahve here the least bad you can think of?

**Comment by markdroth** [comment]
While you're at it, can you also update the following doc to indicate an exception for `[#alpha:]`?

https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md#backwards-compatibility

Thanks!

**Comment by snowp** [comment]
>Is what you ahve here the least bad you can think of?

I think this is the least bad in terms how many people this would be affected by it (only users of the new matching API). I'm also not sure if moving the header matching protos would help without a breaking API change: the protos currently referencing them are not in alpha, so even if we copied them we wouldn't be able to remove the build time dependency on the existing route matching protos since we need to keep the deprecated field around. With this change we can remove the old matching tree proto (immediately? or maybe following a deprecation cycle?) and continue to evolve the new proto without being held back by the API stability policy.

**Comment by markdroth** [comment]
Another thought: Since we're moving this any way, would it make sense to move it to the cncf/xds repo at this point?

**Comment by snowp** [comment]
@markdroth I'm not opposed to that since I don't think it would be too much work, so happy to do that if we feel like it's the right call. @envoyproxy/api-shepherds thoughts?

**Comment by snowp** [comment]
Also: given the exception we're now giving breaking changes to alpha protos, does it still make sense to provide the migration path that this PR offers at the moment? After looking at it some more, none of the affected protos are referenced by type url (only ExtensionWithMatch is, which isnt changed), so if we were to swap the type of the `matcher` field to something that looks the same that should not break anything on the wire

Thoughts?

**Comment by htuch** [comment]
I think moving to `cncf/xds` would make sense as this is a fairly fundamental generic construct. But, don't feel so strongly I'd bike shed a ton. 

Migration path makes sense if it helps out in your situation @snowp, otherwise I think it's fine to just make the breaking change in alpha.

**Comment by snowp** [comment]
Looking into what it would take to migrate this to udpa there are two messages outside of the immediate field that would also need to be migrated: TypedExtensionConfig and matcher.v3.StringMatcher. We could add these to udpa as well, though that will involve duck typing more functions (like I've done in this PR via templating) to make them compatible with the identical protos in the Envoy repo vs the udpa repo. 

Thoughts? Not sure if this makes things more complicated than they have to be

File: api/API_VERSIONING.md [hunk 1]
```diff
@@ -67,6 +67,8 @@ experience a backward compatible break on a change. Specifically:
     may be granted for scenarios in which these stricter conditions model behavior already implied
     structurally or by documentation.
 
+* Messages marked as [#alpha:] are excluded from the backwards compatibility guarantees.
```
Comment (markdroth) [inline api/API_VERSIONING.md:70]: Please put this in the bulletted list below, which lists the other items that have an exception.

**Review by markdroth — COMMENTED** [review]
I'd be fine with creating a copy of both `TypedExtensionConfig` and `StringMatcher` in the cncf/xds repo.

File: api/API_VERSIONING.md [hunk 2]
```diff
@@ -67,11 +67,13 @@ experience a backward compatible break on a change. Specifically:
     may be granted for scenarios in which these stricter conditions model behavior already implied
     structurally or by documentation.
 
+
```
Comment (htuch) [inline api/API_VERSIONING.md:70]: ?

File: api/envoy/extensions/common/matching/v3/extension_matcher.proto [hunk 3]
```diff
@@ -21,8 +24,12 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 //
 // [#alpha:]
 message ExtensionWithMatcher {
+  // The associated matcher. This is deprecated in favor of udpa_matcher.
+  config.common.matcher.v3.Matcher matcher = 1
+      [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
+
   // The associated matcher.
-  config.common.matcher.v3.Matcher matcher = 1 [(validate.rules).message = {required: true}];
+  udpa.type.matcher.v1.Matcher udpa_matcher = 3;
```
Comment (htuch) [inline api/envoy/extensions/common/matching/v3/extension_matcher.proto:32]: This should be `xds.type.matcher..` etc.

File: docs/root/intro/arch_overview/advanced/matching/_include/simple.yaml [hunk 4]
```diff
@@ -29,7 +29,7 @@ static_resources:
                     percentage:
                       numerator: 0
                       denominator: HUNDRED
-              matcher:
+              udpa_matcher:
```
Comment (htuch) [inline docs/root/intro/arch_overview/advanced/matching/_include/simple.yaml:32]: s/udpa/xds/ everywhere (or even just `matcher` is fine)

**Review by htuch — CHANGES_REQUESTED** [review]
/wait

File: docs/root/intro/arch_overview/advanced/matching/_include/simple.yaml (hunk 4 above)
Comment (snowp) [reply docs/root/intro/arch_overview/advanced/matching/_include/simple.yaml:32]: I would do matcher but that's the name of the original field, so I'm having to use something else in order to provide an upgrade path here

**Comment by snowp** [comment]
Seems like the dependency check here is failing since cncf/xds doesn't have any releases? Or perhaps its due to my targeting a non-main commit for now until the cncf/xds PR lands?

**Comment by htuch** [comment]
I think it's lack of any release, but this seems strange it should be the first time since there are other deps in this situation. @moderation @phlax any thoughts?

**Comment by htuch** [comment]
@snowp can you update to `cncf/xds` HEAD and try again? Maybe it had to do with working with branches or something.
/wait

**Comment by htuch** [comment]
Looks like CI failures.
/wait

**Comment by moderation** [comment]
> I think it's lack of any release, but this seems strange it should be the first time since there are other deps in this situation. @moderation @phlax any thoughts?

Seems like the deps issue is resolved and there shouldn't be problems with deps without releases or targeting specific commits.

/lgtm deps

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

**Review by mathetake — APPROVED** [review]
LGTM for changes to SPIFFE validator extension

**Comment by snowp** [comment]
Finally passed CI on this on, ptal @htuch

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**